### PR TITLE
Rewrite rendering engine

### PR DESCRIPTION
**Pull request by numberZero**
*UPD: Note: part of these points may go in other PRs*
* [x] Move to a separate file to reduce compilation time (2 files recompiled on `renderingcore.h` change vs. 28 for `renderingengine.h`)
* [x] Fix 3D rendering
* [ ] Unify 3D rendering methods
* [ ] Fix camera rotation (?)
* [x] Restore undersampling support
* [x] Restore possibility to change 3D mode without full restart
* [ ] Prepare for subsequent changes (shader-based postprocessing)

### CODE REVIEW

File: src/client/renderingcore.cpp [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+RenderingCoreInterlaced::~RenderingCoreInterlaced()
+{
+	clear_textures();
+}
+
+void RenderingCoreInterlaced::init_material()
+{
+	if (mat.MaterialType)
+		return;
```
Comment (numberZero) [inline src/client/renderingcore.cpp]: This is dirty but necessary currently, as smth is NULL at core construction. Really, it should be delayed to game start, when everything is present; that way, settings change wont’t require full restart as well.

File: src/client/renderingcore.cpp [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+	for (int k = 0; k != 3; ++k) {
+		mat.TextureLayer[k].AnisotropicFilter = false;
+		mat.TextureLayer[k].BilinearFilter = false;
+		mat.TextureLayer[k].TrilinearFilter = false;
+		mat.TextureLayer[k].TextureWrapU = irr::video::ETC_CLAMP_TO_EDGE;
+		mat.TextureLayer[k].TextureWrapV = irr::video::ETC_CLAMP_TO_EDGE;
+	}
+}
+
+void RenderingCoreInterlaced::init_textures()
```
Comment (numberZero) [inline src/client/renderingcore.cpp]: yet another `init_textures`... these should be unified

**Comment by numberZero** [comment]
So, the main question: does the code look better this way?

**Comment by nerzhul** [comment]
@numberZero sounds interesting. Can you ensure irrlicht isolation from minetest client ? There are many files which are too near irrlicht, we need to isolate it to prepare a possible future replacement

File: src/client/renderingengine.h [hunk 3]
```diff
@@ -142,44 +141,8 @@ class RenderingEngine
...
-			const irr::core::vector3df &focusPoint, bool show_hud,
-			Camera *camera, Hud *hud, bool draw_wield_tool, Client *client,
-			gui::IGUIEnvironment *guienv, const video::SColor &skycolor);
-
-	video::ITexture *draw_hud(const v2u32 &screensize, bool show_hud, Hud *hud,
-			Client *client, bool draw_crosshair,
-			const video::SColor &skycolor, gui::IGUIEnvironment *guienv,
-			Camera *camera);
-
+	std::unique_ptr<RenderingCore> core;
```
Comment (nerzhul) [inline src/client/renderingengine.h]: nice a c++11 feature :p

**Comment by nerzhul** [comment]
@numberZero i like the way you enhanced it, i never get time to properly cleanup and isolate each included part, it's a nice C++11 code (unique_ptr, override instead of virtual...). 
Do you use clion or clang-tidy to help you to improve those paths ?

File: src/client/renderingengine.h (hunk 3 above)
Comment (Dumbeldor) [inline src/client/renderingengine.h]: Oh yeah ! Good job !

**Comment by numberZero** [comment]
@nerzhul 
> Do you use clion or clang-tidy to help you to improve those paths ?

No.

> Can you ensure irrlicht isolation from minetest client ?

That is, move everything to `RenderingEngine` and co.? That’s what I want; right now there are some weird things like `ClientMap::renderPostFx` which are to be rewritten. That’s a big job, though; I may not have enough time for it.

**Comment by nerzhul** [comment]
we can move things one step by one step, no problem, for me your PR is acceptable if tested in multiple configurations using this form, then we can continue with another change, it's refactor, having a single huge PR is not really easy to approve/review

File: src/client/renderingcore.cpp (hunk 2 above)
Comment (nerzhul) [inline src/client/renderingcore.cpp]: why ? it sounds each mode has its own way

**Comment by rubenwardy** [comment]
Looks like a definite improvement.

Please may you split up the new file pair into a file pair for each class, and move it to maybe client/render

**Comment by numberZero** [comment]
> split up <...> into a file pair for each class

@rubenwardy sure? That’s a lot of files. Also, what to do with `RenderingEngine::_initialize`? I have several ideas but don’t like any too much:
1. Include all the new headers (ugh, 5 more; there are too many already)
2. Use factory approach (but still, where to register these classes? I have a solution, but it looks ugly IMO)

---
The Factory:
```c++
class RenderingCore
{
private:
	// Because static fields have problems with initialization order
	typedef std::map<std::string, RenderingCore * (*)(IrrlichtDevice *)> Constructors;
	static Constructors &getConstructors()
	{
		static Constructors constructors;
		return constructors;
	}

public:
	template <typename Core>
	class Type
	{
		static RenderingCore *create(IrrlichtDevice *device)
		{
			return new Core(device);
		}
	public:
		Type(const std::string &name)
		{
			getConstructors().emplace(name, create);
		}
	};

	static RenderingCore *create(const std::string &type, IrrlichtDevice *device)
	{
		return getConstructors().at(type)(device);
	}
```
Registration:
```c++
static RenderingCore::Type<RenderingCoreAnaglyph> me("anaglyph");
```
This last line looks nice, and it works, but...

**Comment by numberZero** [comment]
@nerzhul @rubenwardy Done. This version seems to work well, with no regression. (undersampling was supported in mono mode only)
Note: I can’t test the `pageflip` mode at all as my hardware lacks the necessary support.

**Comment by nerzhul** [comment]
i look at this tomorrow, it sounds better i just need to check the code accuracy

**Comment by tobyplowy** [comment]
@numberZero this is exiting

You mentioned shader-based postprocessing how far will this allow to go?

**Comment by nerzhul** [comment]
i suggest to merge the PR as is, before any later change, it seems stable, first rewrite engine, second add features, @numberZero @paramat @SmallJoker okay ?

**Comment by numberZero** [comment]
@nerzhul Okay. Faster merging --> faster development, we all want that, I suppose.

**Comment by nerzhul** [comment]
@numberZero other advantage, everybody can benefit from a stable dev and the bisect is easier :)

**Comment by rubenwardy** [comment]
I approve of the architecture, but I haven't tested

**Comment by rubenwardy** [comment]
after pressing f2 or f1 in anaglyph mode it looks like the background isn't clearing:



That's flickering when I stay still, and when I rotate the sun gives off a tail.

--------

top bottom seg faults

```
2017-08-26 02:25:50: WARNING[Main]: NodeDefManager: Ignoring CONTENT_IGNORE redefinition
[New Thread 0x7fffd28b6700 (LWP 19567)]
2017-08-26 02:25:50: ACTION[Main]:         .__               __                   __   
2017-08-26 02:25:50: ACTION[Main]:   _____ |__| ____   _____/  |_  ____   _______/  |_ 
2017-08-26 02:25:50: ACTION[Main]:  /     \|  |/    \_/ __ \   __\/ __ \ /  ___/\   __\
2017-08-26 02:25:50: ACTION[Main]: |  Y Y  \  |   |  \  ___/|  | \  ___/ \___ \  |  |  
2017-08-26 02:25:50: ACTION[Main]: |__|_|  /__|___|  /\___  >__|  \___  >____  > |__|  
2017-08-26 02:25:50: ACTION[Main]:       \/        \/     \/          \/     \/        
2017-08-26 02:25:50: ACTION[Main]: World at [/home/ruben/.minetest/worlds/aaaaaaaa]
2017-08-26 02:25:50: ACTION[Main]: Server for gameid="minetest" listening on 0.0.0.0:52458.
[New Thread 0x7fffd38cf700 (LWP 19568)]
[New Thread 0x7fffd30ce700 (LWP 19569)]
[New Thread 0x7fffbbfff700 (LWP 19570)]
[New Thread 0x7fffbae69700 (LWP 19573)]
2017-08-26 02:25:51: ACTION[Server]: singleplayer [127.0.0.1] joins game. 
2017-08-26 02:25:51: ACTION[Server]: singleplayer joins game. List of players: singleplayer
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Invalid rendering mode: topbottom

Thread 1 "minetest" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:58
58	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt full
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:58
        set = {__val = {4096, 140737353949760, 140737488339632, 8589918864, 140737353948904, 140737488339616, 140737307415835, 252833149, 4294967295, 140737353951424, 140737297931736, 140737353936896, 140737488339712, 4294967296, 140737353948904, 140737488339696}}
        pid = <optimised out>
        tid = <optimised out>
#1  0x00007ffff4a9637a in __GI_abort () at abort.c:89
        save_stage = 2
        act = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {0, 140737301845280, 3432, 1, 140737301845411, 0, 140737298412594, 11, 140737310912360, 140737353828160, 140737353828160, 1, 140737353830736, 93825110674256, 
              140737301849936, 140737353828416}}, sa_flags = -180534048, sa_restorer = 0x55555c622028}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00007ffff53d756d in __gnu_cxx::__verbose_terminate_handler() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#3  0x00007ffff53d5316 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#4  0x00007ffff53d5361 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#5  0x00007ffff53d5579 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#6  0x0000555555b45dc6 in createRenderingCore (stereo_mode="topbottom", device=0x5555565cfdd0, client=0x555557cb5040, hud=0x55555c5975e0) at /home/ruben/dev/minetest/src/client/render/factory.cpp:41
No locals.
#7  0x0000555555b4ba25 in RenderingEngine::_initialize (this=0x5555565d2200, client=0x555557cb5040, hud=0x55555c5975e0) at /home/ruben/dev/minetest/src/client/renderingengine.cpp:443
        draw_mode = "topbottom"
#8  0x0000555555de3793 in RenderingEngine::initialize (client=0x555557cb5040, hud=0x55555c5975e0) at /home/ruben/dev/minetest/src/client/renderingengine.h:122
No locals.
#9  0x0000555555dcea4c in Game::startup (this=0x7fffffffc890, kill=0x555556573b60 <porting::g_killed>, random_input=false, input=0x555556b1fd90, map_dir="/home/ruben/.minetest/worlds/aaaaaaaa", playername="singleplayer", password="", address=0x7fffffffc7b0, 
    port=52458, error_message="", reconnect=0x7fffffffd674, chat_backend=0x7fffffffd860, gamespec=..., simple_singleplayer_mode=true) at /home/ruben/dev/minetest/src/game.cpp:1631
No locals.
#10 0x0000555555de2459 in the_game (kill=0x555556573b60 <porting::g_killed>, random_input=false, input=0x555556b1fd90, map_dir="/home/ruben/.minetest/worlds/aaaaaaaa", playername="singleplayer", password="", address="", port=52458, error_message="", 
    chat_backend=..., reconnect_requested=0x7fffffffd674, gamespec=..., simple_singleplayer_mode=true) at /home/ruben/dev/minetest/src/game.cpp:4744
        game = {input = 0x555556b1fd90, client = 0x555557cb5040, server = 0x555556bcc870, texture_src = 0x555556bd2c40, shader_src = 0x555556bd6510, itemdef_manager = 0x555556d2ae30, nodedef_manager = 0x555556d91d50, soundfetcher = {<OnDemandSoundFetcher> = {
              _vptr.OnDemandSoundFetcher = 0x555556512678 <vtable for GameOnDemandSoundFetcher+16>}, m_fetched = std::set with 0 elements}, sound = 0x555556bc6650, sound_is_dummy = false, soundmaker = 0x555556c97da0, chat_backend = 0x7fffffffd860, 
          current_formspec = 0x0, cur_formname = "", eventmgr = 0x555556bb6bf0, quicktune = 0x555556bc6600, gui_chat_console = 0x55555c50b330, draw_control = 0x555556c01fb0, camera = 0x55555c44fae0, clouds = 0x55555c49e7c0, sky = 0x55555c584d90, 
          local_inventory = 0x55555c3a0740, hud = 0x55555c5975e0, mapper = 0x555556c452f0, runData = {dig_index = 0, new_playeritem = 0, pointed_old = {type = POINTEDTHING_NOTHING, node_undersurface = {X = 0, Y = 0, Z = 0}, node_abovesurface = {X = 0, Y = 0, 
                Z = 0}, node_real_undersurface = {X = 0, Y = 0, Z = 0}, object_id = 0, intersection_point = {X = 0, Y = 0, Z = 0}, intersection_normal = {X = 0, Y = 0, Z = 0}, distanceSq = 0}, digging = false, ldown_for_dig = false, dig_instantly = false, 
            digging_blocked = false, left_punch = false, update_wielded_item_trigger = true, reset_jump_timer = false, nodig_delay_timer = 0, dig_time = 0, dig_time_complete = 0, repeat_rightclick_timer = 0, object_hit_delay_timer = 0, time_from_last_punch = 10, 
            selected_object = 0x0, jump_timer = 0, damage_flash = 0, update_draw_list_timer = 0, statustext_time = 0, fog_range = 0, update_draw_list_last_cam_dir = {X = 0, Y = 0, Z = 0}, profiler_current_page = 0, profiler_max_page = 3, time_of_day = 0, 
            time_of_day_smooth = 0}, flags = {show_chat = true, show_hud = true, show_minimap = false, force_fog_off = false, show_debug = true, show_profiler_graph = false, disable_camera_update = false}, device = 0x5555565cfdd0, driver = 0x5555568b2840, 
          smgr = 0x555556b1db70, kill = 0x555556573b60 <porting::g_killed>, error_message = 0x7fffffffd6c0, reconnect_requested = 0x7fffffffd674, skybox = 0x0, random_input = false, simple_singleplayer_mode = true, crack_animation_length = 5, 
          guitext = 0x55555c4f6060, guitext2 = 0x55555c47b090, guitext3 = 0x55555c490cb0, guitext_info = 0x55555c4ed850, guitext_status = 0x55555c5831a0, guitext_chat = 0x55555c4fa5c0, guitext_profiler = 0x55555c596f00, infotext = L"", m_statustext = L"", 
          keycache = {key = {{Key = irr::KEY_KEY_W, Char = 87 L'W', m_name = "KEY_KEY_W"}, {Key = irr::KEY_KEY_S, Char = 83 L'S', m_name = "KEY_KEY_S"}, {Key = irr::KEY_KEY_A, Char = 65 L'A', m_name = "KEY_KEY_A"}, {Key = irr::KEY_KEY_D, Char = 68 L'D', 
                m_name = "KEY_KEY_D"}, {Key = irr::KEY_SPACE, Char = 0 L'\000', m_name = "KEY_SPACE"}, {Key = irr::KEY_KEY_E, Char = 69 L'E', m_name = "KEY_KEY_E"}, {Key = irr::KEY_LSHIFT, Char = 0 L'\000', m_name = "KEY_LSHIFT"}, {Key = irr::KEY_KEY_CODES_COUNT, 
                Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_ESCAPE, Char = 0 L'\000', m_name = "KEY_ESCAPE"}, {Key = irr::KEY_KEY_Q, Char = 81 L'Q', m_name = "KEY_KEY_Q"}, {Key = irr::KEY_KEY_I, Char = 73 L'I', m_name = "KEY_KEY_I"}, {Key = irr::KEY_KEY_T, 
                Char = 84 L'T', m_name = "KEY_KEY_T"}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 47 L'/', m_name = "/"}, {Key = irr::KEY_PERIOD, Char = 46 L'.', m_name = "KEY_PERIOD"}, {Key = irr::KEY_F10, Char = 0 L'\000', m_name = "KEY_F10"}, {Key = irr::KEY_F9, 
                Char = 0 L'\000', m_name = "KEY_F9"}, {Key = irr::KEY_KEY_K, Char = 75 L'K', m_name = "KEY_KEY_K"}, {Key = irr::KEY_KEY_J, Char = 74 L'J', m_name = "KEY_KEY_J"}, {Key = irr::KEY_KEY_H, Char = 72 L'H', m_name = "KEY_KEY_H"}, {Key = irr::KEY_KEY_B, 
                Char = 66 L'B', m_name = "KEY_KEY_B"}, {Key = irr::KEY_KEY_N, Char = 78 L'N', m_name = "KEY_KEY_N"}, {Key = irr::KEY_KEY_M, Char = 77 L'M', m_name = "KEY_KEY_M"}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {
                Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_F12, Char = 0 L'\000', m_name = "KEY_F12"}, {Key = irr::KEY_F1, Char = 0 L'\000', m_name = "KEY_F1"}, {
                Key = irr::KEY_F2, Char = 0 L'\000', m_name = "KEY_F2"}, {Key = irr::KEY_F3, Char = 0 L'\000', m_name = "KEY_F3"}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_F5, Char = 0 L'\000', m_name = "KEY_F5"}, {
                Key = irr::KEY_F6, Char = 0 L'\000', m_name = "KEY_F6"}, {Key = irr::KEY_F7, Char = 0 L'\000', m_name = "KEY_F7"}, {Key = irr::KEY_PLUS, Char = 43 L'+', m_name = "KEY_PLUS"}, {Key = irr::KEY_MINUS, Char = 45 L'-', m_name = "KEY_MINUS"}, {
                Key = irr::KEY_KEY_R, Char = 82 L'R', m_name = "KEY_KEY_R"}, {Key = irr::KEY_KEY_Z, Char = 90 L'Z', m_name = "KEY_KEY_Z"}, {Key = irr::KEY_END, Char = 0 L'\000', m_name = "KEY_END"}, {Key = irr::KEY_HOME, Char = 0 L'\000', m_name = "KEY_HOME"}, {
                Key = irr::KEY_PRIOR, Char = 0 L'\000', m_name = "KEY_PRIOR"}, {Key = irr::KEY_NEXT, Char = 0 L'\000', m_name = "KEY_NEXT"}, {Key = irr::KEY_KEY_P, Char = 80 L'P', m_name = "KEY_KEY_P"}, {Key = irr::KEY_KEY_1, Char = 49 L'1', 
                m_name = "KEY_KEY_1"}, {Key = irr::KEY_KEY_2, Char = 50 L'2', m_name = "KEY_KEY_2"}, {Key = irr::KEY_KEY_3, Char = 51 L'3', m_name = "KEY_KEY_3"}, {Key = irr::KEY_KEY_4, Char = 52 L'4', m_name = "KEY_KEY_4"}, {Key = irr::KEY_KEY_5, Char = 53 L'5', 
                m_name = "KEY_KEY_5"}, {Key = irr::KEY_KEY_6, Char = 54 L'6', m_name = "KEY_KEY_6"}, {Key = irr::KEY_KEY_7, Char = 55 L'7', m_name = "KEY_KEY_7"}, {Key = irr::KEY_KEY_8, Char = 56 L'8', m_name = "KEY_KEY_8"}, {Key = irr::KEY_KEY_9, Char = 57 L'9', 
                m_name = "KEY_KEY_9"}, {Key = irr::KEY_KEY_0, Char = 48 L'0', m_name = "KEY_KEY_0"}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, 
                Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {
                Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', 
                m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, 
                Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}, {Key = irr::KEY_KEY_CODES_COUNT, Char = 0 L'\000', m_name = ""}}, 
            handler = 0x555556b1fd90}, profiler_interval = {m_accumulator = 0}, m_cache_doubletap_jump = false, m_cache_enable_clouds = true, m_cache_enable_joysticks = false, m_cache_enable_particles = true, m_cache_enable_fog = true, 
          m_cache_enable_noclip = false, m_cache_enable_free_move = false, m_cache_mouse_sensitivity = 0.200000003, m_cache_joystick_frustum_sensitivity = 170, m_repeat_right_click_time = 0.25, m_cache_cam_smoothing = 1, m_cache_fog_start = 0.400000006, 
          m_invert_mouse = false, m_first_loop_after_window_activation = true, m_camera_offset_changed = false}
        server_address = ""
#11 0x0000555555b33df8 in ClientLauncher::run (this=0x7fffffffdd60, game_params=..., cmd_args=...) at /home/ruben/dev/minetest/src/client/clientlauncher.cpp:236
        game_has_run = true
        text = 0x555556b5e960 L"\x61616161\x61616161\x72617000\x74656d61\x205d7265\x77656e5b\x7364005d\x726f772f摬"
        video_driver = 0x5555568b2840
        __PRETTY_FUNCTION__ = "bool ClientLauncher::run(GameParams&, const Settings&)"
        camera = 0x555556b5e3f0
        chat_backend = {m_console_buffer = {m_scrollback = 500, m_unformatted = std::vector of length 0, capacity 0, m_cols = 0, m_rows = 0, m_scroll = 0, m_formatted = std::vector of length 0, capacity 0, m_empty_formatted_line = {
              fragments = std::vector of length 0, capacity 0, first = true}}, m_recent_buffer = {m_scrollback = 6, m_unformatted = std::vector of length 0, capacity 0, m_cols = 0, m_rows = 0, m_scroll = 0, m_formatted = std::vector of length 0, capacity 0, 
            m_empty_formatted_line = {fragments = std::vector of length 0, capacity 0, first = true}}, m_prompt = {m_prompt = L"]", m_line = L"", m_history = std::vector of length 0, capacity 0, m_history_index = 0, m_history_limit = 500, m_cols = 0, m_view = 0, 
---Type <return> to continue, or q <return> to quit---
            m_cursor = 0, m_cursor_len = 0, m_nick_completion_start = 0, m_nick_completion_end = 0}}
        error_message = ""
        reconnect_requested = false
        first_loop = false
        retval = true
        kill = 0x555556573b60 <porting::g_killed>
#12 0x0000555555e8f051 in main (argc=1, argv=0x7fffffffe0a8) at /home/ruben/dev/minetest/src/main.cpp:209
        retval = 32767
        cmd_args = {m_settings = std::unordered_map with 0 elements, m_defaults = std::unordered_map with 0 elements, m_callbacks = std::unordered_map with 0 elements, m_callback_mutex = {<std::__mutex_base> = {_M_mutex = {__data = {__lock = 0, __count = 0, 
                  __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}}, <No data fields>}, m_mutex = {<std::__mutex_base> = {_M_mutex = {__data = {__lock = 0, 
                  __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}}, <No data fields>}}
        cmd_args_ok = true
        __debug_stacker = {m_stack = 0x5555565bf5e0, m_overflowed = false}
        __PRETTY_FUNCTION__ = "int main(int, char**)"
        game_params = {socket_port = 30000, world_path = "/home/ruben/.minetest/worlds/world", game_spec = {id = "minetest", path = "/home/ruben/dev/minetest/bin/../games/minetest_game", gamemods_path = "/home/ruben/dev/minetest/bin/../games/minetest_game/mods", 
            addon_mods_paths = std::set with 2 elements = {[0] = "/home/ruben/.minetest/mods", [1] = "/home/ruben/dev/minetest/bin/../mods"}, name = "Minetest Game", menuicon_path = "/home/ruben/dev/minetest/bin/../games/minetest_game/menu/icon.png"}, 
          is_dedicated_server = false}
        isServer = false
        launcher = {list_video_modes = false, skip_main_menu = false, use_freetype = true, random_input = false, address = "", playername = "", password = "", input = 0x555556b1fd90, receiver = 0x5555565d23e0, skin = 0x555556b1d8b0, font = 0x0, gamespec = {
            id = "minetest", path = "/home/ruben/dev/minetest/bin/../games/minetest_game", gamemods_path = "/home/ruben/dev/minetest/bin/../games/minetest_game/mods", addon_mods_paths = std::set with 2 elements = {[0] = "/home/ruben/.minetest/mods", 
              [1] = "/home/ruben/dev/minetest/bin/../mods"}, name = "Minetest Game", menuicon_path = "/home/ruben/dev/minetest/bin/../games/minetest_game/menu/icon.png"}, worldspec = {path = "/home/ruben/.minetest/worlds/aaaaaaaa", name = "aaaaaaaa", 
            gameid = "minetest"}, simple_singleplayer_mode = true, current_playername = "singleplayer", current_password = "", current_address = "", current_port = 52458}
```

---------

side by side works, but everything is stretched:

**Comment by numberZero** [comment]
@rubenwardy Thanks for the feedback.
1. Fixed anaglyph mode, although I don’t understand why could it behave that way. Also fixed a problem with the pause menu;
2. In the `topbottom` mode, there was not a segfault (`SIGSEGV`) but an uncaught exception (raising `SIGABRT`) as the mode was not known. I fixed that as well.
3. On side-by-side and top-bottom modes, I have 2 questions:
   1. Should the image be stretched (as currently) or not?
   2. Should the GUI be present on both sides (as currently), or be drawn over the whole screen?

**Comment by rubenwardy** [comment]
> Should the image be stretched (as currently) or not?

No. If the real screen size is 1000 x 500 then the screen and GUI should be drawn at 500 x 500 each. Ie: no stretching.

> Should the GUI be present on both sides (as currently), or be drawn over the whole screen?

On each. Ideally they should have a configurable offset, as to appear like it's paper floating in space. This wasn't done previously, however - so it's fine to leave it as-is

**Comment by numberZero** [comment]
That appeared to be more problematic than I thought, but using 2 hacks
I managed with that.
BTW, am I right that stereo modes are non-functional in `master`?

**Comment by nerzhul** [comment]
@numberZero can you take time to finish this PR please ?

**Comment by numberZero** [comment]
@nerzhul I thought it is ready. Perhaps I didn’t notice that it fails lint. Are there any other problems?

**Comment by nerzhul** [comment]
can you fix lint ?

File: src/client/render/interlaced.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+void RenderingCoreInterlaced::initMaterial()
+{
+	IShaderSource *s = client->getShaderSource();
+	mat.UseMipMaps = false;
+	mat.ZBuffer = false;
+	mat.ZWriteEnable = false;
+	u32 shader = s->getShader("3d_interlaced_merge", TILE_MATERIAL_BASIC, 0);
+	mat.MaterialType = s->getShaderInfo(shader).material;
+	for (int k = 0; k != 3; ++k) {
```
Comment (nerzhul) [inline src/client/render/interlaced.cpp]: k < 3

File: src/client/render/interlaced.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+{
+	driver->removeTexture(left);
+	driver->removeTexture(right);
+	driver->removeTexture(mask);
+}
+
+void RenderingCoreInterlaced::initMask()
+{
+	u8 *data = reinterpret_cast<u8 *>(mask->lock());
+	for (unsigned j = 0; j < screensize.Y; j++) {
```
Comment (nerzhul) [inline src/client/render/interlaced.cpp]: u32

File: src/client/render/plain.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "plain.h"
+#include "settings.h"
+
+inline unsigned scaledown(unsigned coef, unsigned size)
```
Comment (nerzhul) [inline src/client/render/plain.cpp]: u32

**Comment by SmallJoker** [comment]
Testing results:
- anaglyph
  - nothing suspicious found
- interlaced (testing on an LCD)
  - text, formspecs and HUD is shown, whereas no mapblocks are visible (intended behaviour?)
- topbottom
  - mapblocks near window border are not rendered (wrong fov calculation?)
- sidebyside
  - Pause menu centred wrong (inventory and key/volume change OK)
  - Status messages like "free_mode enabled" centred wrong
- pageflip (LCD, NVIDIA Quadro FX 4600)
  - black screen where the mapblocks should be shown (intended behaviour?)

**Comment by sapier** [comment]
really good work!
one small thing useEye() uses true/false as calling parameters. I know there are only two options (right now) but it makes code hard to read as neither true nor false are at any way connected to Right or left.
What about Enum Right/Left?

**Comment by sapier** [comment]
I just managed to get the first 3d-vr images within minetest. Atm it looks like abusing your great new design as for vr (as I did it by now) the drawing mechanics is completely different. 
I try to find a way doing it more consistent to the new design the next days. Maybe we should talk about what you prefere once I have some suggestions.

**Comment by numberZero** [comment]
@nerzhul Fixed.
@SmallJoker 
*    interlaced — no mapblocks are visible
     * Requires shaders (updated docs to mention that)
*    sidebyside — Pause menu centred wrong
     * IrrLicht does it that way when the area available is too small.
*    pageflip
      * Can’t test, my system doesn’t support that at all.
*    topbottom — mapblocks near window border are not rendered
*    sidebyside — Status messages centred wrong
     * Should be fixable...

**Comment by vladtcvs** [comment]
Anaglyph mode requires color correction. Most of the minetest surface is a green or blue. So in blue glass worlds looks pretty similar to the original world, and in red glass it looks dark-grey.

**Comment by numberZero** [comment]
@SmallJoker so..?
If you could say that pageflip mode was working and now it doesn’t, there would be a problem. Otherwise, unless you can fix it yourself, it may be better to leave it as-is for now.

@vladtcvs Thanks for the feedback. That certainly won’t be a part of this PR, but may be made later. So, should the correction be applied before or after red-cyan split?

**Comment by numberZero** [comment]
And well, @nerzhul, I [fixed](https://github.com/minetest/minetest/pull/6253/commits/db90f3873aac9e33f503fbaa8e36253ff2712b7d) [these points](https://github.com/minetest/minetest/pull/6253#pullrequestreview-67830471) some time ago. It even passes LINT =)

**Comment by nerzhul** [comment]
nice job thanks

**Comment by alwayshopeless** [comment]
Reviving discus. It is better to start post-processing by switching to a new version of GLSL, this will make it easier to port existing shaders and it may be worth taking a look at the structure of shaders used in other popular games so that they can be easily ported from there.

**Comment by numberZero** [comment]
> Reviving discus. It is better to start post-processing by switching to a new version of GLSL, this will make it easier to port existing shaders and it may be worth taking a look at the structure of shaders used in other popular games so that they can be easily ported from there.

Um, what? After 6 years I don’t exactly remember what this PR was all about.

**Pull request merged**